AK/BigIntBase.h

```diff
@@ -0,0 +1,595 @@
+
+    constexpr static bool is_signed = is_signed_;
+
+    explicit constexpr StorageSpan(Span<Word> span)
+        : Span<Word>(span)
+    {
+    }
+
+    constexpr bool is_negative() const
+    {
+        return is_signed && this->last() >> (word_size - 1);
+    }
+};
+
+using UnsignedStorageSpan = StorageSpan<NativeWord, false>;
+using UnsignedStorageReadonlySpan = StorageSpan<NativeWord const, false>;
+
+// Sometimes we want to know the exact maximum amount of the bits required to represent the number.
+// However, the bit size only acts as a hint for wide multiply operations. For all other purposes,
+// `bit_size`-sized and `ceil(bit_size / word_size) * word_size`-sized `StaticStorage`s will act the
+// same.
+template<bool is_signed_, size_t bit_size>
+requires(bit_size <= max_big_int_length * word_size) struct StaticStorage {
+    constexpr static size_t static_size = (bit_size + word_size - 1) / word_size;
+    constexpr static bool is_signed = is_signed_;
+
+    // We store integers in little-endian regardless of the host endianness. We use two's complement
+    // representation of negative numbers and do not bother at all if `bit_size % word_size != 0`
+    // (i. e. do not properly handle overflows).
+    NativeWord m_data[static_size];
+
+    constexpr bool is_negative() const
+    {
+        return is_signed_ && m_data[static_size - 1] >> (word_size - 1);
+    }
+
+    constexpr static size_t size()
+    {
+        return static_size;
+    }
+
+    constexpr NativeWord operator[](size_t i) const
+    {
+        return m_data[i];
+    }
+
+    constexpr NativeWord& operator[](size_t i)
+    {
+        return m_data[i];
+    }
+
+    constexpr NativeWord const* data() const
+    {
+        return m_data;
+    }
+
+    constexpr NativeWord* data()
+    {
+        return m_data;
+    }
+};
+
+// There is no reason to ban u128{0} + 1 (although the second argument type is signed, the value is
+// known at the compile time to be non-negative). In order to do so, we provide overloads in
+// UFixedBigInt which take IntegerWrapper as an argument.
+struct IntegerWrapper {
+    StaticStorage<false, bit_width<int>> m_data;
+
+    consteval IntegerWrapper(int value)
+    {
+        if (value < 0)
+            (new int)&&"Requested implicit conversion of an integer to the unsigned one will underflow.";
+        m_data[0] = static_cast<NativeWord>(value);
+    }
+};
+
+constexpr inline auto get_storage_of(IntegerWrapper value) { return value.m_data; }
+
+template<BuiltInUFixedInt T>
+constexpr StaticStorage<false, bit_width<T>> get_storage_of(T value)
+{
+    if constexpr (sizeof(T) > sizeof(NativeWord)) {
+        static_assert(sizeof(T) == 2 * sizeof(NativeWord));
+        return { static_cast<NativeWord>(value), static_cast<NativeWord>(value >> word_size) };
+    }
+    return { static_cast<NativeWord>(value) };
+}
+
+// ===== Utilities =====
+template<typename T>
+ALWAYS_INLINE constexpr void taint_for_optimizer(T& value)
```

Comment:
Used. https://github.com/DanShaders/serenity/blob/999c58395e0cde03ae0c8a5e3b86e6564eb60433/AK/UFixedBigInt.h#L480 And the intent is to avoid leaking information about encryption keys somewhere in LibCrypto.